cozy/structures/heaps.py

```diff
@@ -66,6 +66,34 @@ def heap_func(e : Exp, concretization_functions : { str : Exp } = None) -> ELamb
         return ELambda(v, ECond(e.cond, h1.apply_to(v), h2.apply_to(v)).with_type(h1.body.type))
     raise NotImplementedError(repr(e))
 
+
+def heap_elems(e : Exp, concretization_functions : { str : Exp } = None) -> Exp:
+    """
+    Assuming 'e' produces a heap, this returns its elements.
+    """
+    if isinstance(e, EMakeMinHeap) or isinstance(e, EMakeMaxHeap):
+        return e.e
+    if isinstance(e, EVar):
+        if concretization_functions:
+            ee = concretization_functions.get(e.id)
+            if ee is not None:
+                return heap_elems(ee)
+        else:
+            tt = TTuple((e.type.elem_type, e.type.key_type))
+            return EMap(e, mk_lambda(tt, lambda arg: ETupleGet(arg, 0).with_type(e.type.elem_type))).with_type(e.type)
```

Comment:
This EMap expression does not seem right. This code will work within the solver and evaluator because of how `encode` is defined. However, Cozy will generate incorrect code if this ever leaks out to generated code, due to how `rep_type` is defined. The right thing to do here is to throw an exception; there is not enough information to return a correct answer. If we ever see that exception being thrown, we will need to re-think what the caller is doing.

Reply:
`storage_size` will be called with `EVar('x').with_type(TMinHeap(INT, INT))`, and failed in `heap_elems` if we don't do "encode" here

Reply:
Do you have an example that demonstrates the misbehavior? If it is doing this, then there is a problem somewhere in the cost model I suspect. (I don't see this line being reached when I run `cozy examples/maxbag.ds --java examples/MaxBag.java`.)

Reply:
> (I don't see this line being reached when I run cozy examples/maxbag.ds --java examples/MaxBag.java.

You mean in the current diff? You need to revert part of my code in this PR, making `concretization_functions` necessary, similar to `heap_func`. Then run `cozy examples/maxbag.ds --java examples/MaxBag.java`. I can observe the crash here.